go-controller/pkg/node/node.go

```diff
@@ -469,10 +475,6 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
 						klog.Fatalf("Failed to upgrade service route for node, error: %v", err)
 					}
 				}
-				// ensure CNI support for port binding built into OVN, as masters have been upgraded
```

Comment:
why remove this?

Reply:
This function enables ovn-installed usage based on ovn-kube version, but it should consider ovn version, which is checked here https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/node.go#L251, and this flag is passed to NewCNIServer, and from that point it shouldn't change.

Reply:
@trozet  does that make sense?

Reply:
So the check you referenced on the CNI side does only check during ovnkube node start time, and its checking an entry in sbdb: 
```
	if _, stderr, err := util.RunOVNSbctl("--columns=up", "list", "Port_Binding"); err != nil {
		if strings.Contains(stderr, "does not contain a column") {
			klog.Infof("Falling back to using legacy CNI OVS flow readiness checks")
			return false, nil
		}
```

However, during upgrade ovnkube-node and ovn-controller will upgrade first. That means ovnkube-node will start and check SBDB which is not upgraded and does not have this column and think it does not support ovn-installed. Then when master upgrades, it will update the topo version and the node will read it here and enable the support.

I think this code needs to stay. However you point out that we assume if you are running this version of ovn-kubernetes then you have an OVN that is new enough to support ovn-installed. To fix that you can simply add ```getOVNIfUpCheckMode() ``` here as well. So:

```
if initialTopoVersion < types.OvnPortBindingTopoVersion && cniServer != nil  && getOVNIfUpCheckMode() {
```

Reply:
Ok now I get it, upgrade order is not obvious, thanks for explanation!